### PR TITLE
[feature] Add a BitSet implementation based on RoaringBitmap

### PR DESCRIPTION
**Pull request by dao-jun**
### SUMMARY

Add a `BitSet` implementation based on RoaringBitmap to make users could migrate `BitSet` to `RoaringBitmap` easier.

### Automated Checks

- [x] I have run `./gradlew test` and made sure that my PR does not break any unit test.
- [x] I have run `./gradlew checkstyleMain` or the equivalent and corrected the formatting warnings reported.

### CODE REVIEW

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  @Override
+  public boolean equals(Object obj) {
+    if (obj == this) {
+      return true;
+    }
+    if (obj instanceof RoaringBitSet) {
+      return roaringBitmap.equals(((RoaringBitSet) obj).roaringBitmap);
+    }
+    if (obj instanceof BitSet) {
+      return roaringBitmap.equals(fromBitSet((BitSet) obj));
```
Comment (szarnekow) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java:188]: Do there produce the same hashCode?

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    }
+    return false;
+  }
+
+  @Override
+  public Object clone() {
+    super.clone();
```
Comment (szarnekow) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java:195]: Probably not what you want to do here.

Comment (szarnekow) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java:195]: Not that I'm not a committer and the fact that I read through the PR should not indicate that the feature is or is not considered to be a valid request on this project.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java (hunk 1 above)
Comment (dao-jun) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java]: It does not produce the same hashcode, thanks for reminder, fixed

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java (hunk 2 above)
Comment (dao-jun) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java]: fixed

Comment (dao-jun) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java]: > Not that I'm not a committer and the fact that I read through the PR should not indicate that the feature is or is not considered to be a valid request on this project.

Thanks for you explain, it doesn't matter that whether this project accept or reject this PR.
 I did the feature in another project, but it's great that if the upstream project could accept it.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+  @Override
+  public String toString() {
+    return roaringBitmap.toString();
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
+  public void flip(int bitIndex) {
+    roaringBitmap.flip(bitIndex, bitIndex + 1);
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java:209]:
```suggestion
    roaringBitmap.flip((long)bitIndex, (long)bitIndex + 1);
```

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
+  public void flip(int fromIndex, int toIndex) {
+    roaringBitmap.flip(fromIndex, toIndex);
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java:215]:
```suggestion
    roaringBitmap.flip((long)fromIndex, (long)toIndex);
```

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  @SuppressWarnings("deprecation")
+  public void set(int fromIndex, int toIndex) {
+    roaringBitmap.add(fromIndex, toIndex);
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
+  public void set(int fromIndex, int toIndex, boolean value) {
+    if (value) {
+      roaringBitmap.add(fromIndex, toIndex);
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java:48]:
```suggestion
      roaringBitmap.add((long)fromIndex, (long)toIndex);
```

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+  @Override
+  public void clear(int bitIndex) {
+    roaringBitmap.remove(bitIndex);
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
+  public void clear(int fromIndex, int toIndex) {
+    roaringBitmap.remove(fromIndex, toIndex);
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitSet.java:62]:
```suggestion
    roaringBitmap.remove((long)fromIndex, (long)toIndex);
```

File: RoaringBitmap/src/test/java/org/roaringbitmap/RoaringBitSetTest.java [hunk 7]
```diff
@@ -0,0 +1,912 @@
...
+  @Test
+  public void testAndNot() {
+    int failCount = 0;
+
+    for (int i = 0; i < 100; i++) {
+      RoaringBitSet b1 = new RoaringBitSet();
+      RoaringBitSet b2 = new RoaringBitSet();
+
+      // Set some random bits in first set and remember them
+      int nextBitToSet = 0;
```
Comment (lemire) [inline RoaringBitmap/src/test/java/org/roaringbitmap/RoaringBitSetTest.java:299]:
```suggestion
```

File: RoaringBitmap/src/test/java/org/roaringbitmap/RoaringBitSetTest.java [hunk 8]
```diff
@@ -0,0 +1,912 @@
...
+  @Test
+  public void testAnd() {
+    int failCount = 0;
+
+    for (int i = 0; i < 100; i++) {
+      RoaringBitSet b1 = new RoaringBitSet();
+      RoaringBitSet b2 = new RoaringBitSet();
+
+      // Set some random bits in first set and remember them
+      int nextBitToSet = 0;
```
Comment (lemire) [inline RoaringBitmap/src/test/java/org/roaringbitmap/RoaringBitSetTest.java:333]:
```suggestion
```

**Comment by lemire** [comment]
@dao-jun Please have a look at my proposals.

**Comment by dao-jun** [comment]
@lemire Thanks for your review, addressed

**Comment by lemire** [comment]
Running tests.

**Comment by lemire** [comment]
I like it, but I'd love to have feedback from others. Note that the instances will have additional dead-weight overhead because you need to carry the BitSet class around... but I am not sure how of a concern that is.

@richardstartin : any opinion?

Anyone else?

**Review by lemire — APPROVED** [review]
I approve but I'd like to get more feedback before merging this.

**Comment by dao-jun** [comment]
> I like it, but I'd love to have feedback from others. Note that the instances will have additional dead-weight overhead because you need to carry the BitSet class around... but I am not sure how of a concern that is.
> 
> @richardstartin : any opinion?
> 
> Anyone else?

Yes, RoaringBitSet extends BitSet, one single RoaringBitSet instance may consume extra serval bytes, but it also comes one benefit: users *almost* takes no changes when migrate BitSet to RoaringBitSet.

Anyway, it's a trade, extends BitSet or not do have its reason.

**Comment by richardstartin** [comment]
no opinion, it's an imperfect solution because of the constraints, but it's useful.

**Comment by lemire** [comment]
I am going to merge it. Thanks @richardstartin

**Comment by dao-jun** [comment]
@lemire Hi, I don't find the 1.1.0 release in maven central repo, it there something goes wrong?

**Comment by lemire** [comment]
(Note that releases do not appear immediately. The maven repository is quite slow. It often takes a few days if not more.)

**Comment by dao-jun** [comment]
I was just wandering if the dist published to maven central repo.

Seems it's maven's issue, thanks for your explain, I got it.

**Comment by lemire** [comment]
Yes, our releases are uploaded to the maven repository using sonatype tooling.

**Comment by lhotari** [comment]
I made a follow-up PR #727 to properly implement `RoaringBitSet.get(int fromIndex, int toIndex)` without relying on `java.util.BitSet` implementation for `get(int fromIndex, int toIndex)`. Please review!

**Pull request merged**